torchtune/__init__.py

```diff
@@ -3,40 +3,96 @@
 #
 # This source code is licensed under the BSD-style license found in the
 # LICENSE file in the root directory of this source tree.
+from dataclasses import dataclass
+from typing import List
 
 from torchtune import datasets, models, modules, utils
 
-_RECIPE_LIST = [
-    "full_finetune_single_device.py",
-    "full_finetune_distributed.py",
-    "alpaca_generate.py",
-    "lora_finetune_single_device.py",
-    "lora_finetune_distributed.py",
-    "eleuther_eval.py",
+
+@dataclass
+class Config:
+    name: str
+    file_path: str
+
+
+@dataclass
+class Recipe:
+    name: str
+    file_path: str
+    configs: List[Config]
+    supports_distributed: bool
+
+    def get_configs(self) -> List[Config]:
+        return self.configs
+
+
+_ALL_RECIPES = [
```

Comment:
Yeh ```__init__.py``` is the last place I'd look for this. Why not just define this in a recipes.py or something like that and then just import ALL_RECIPES? Also it's a big confusing that this is a private dict in ```__init__.py```.

Reply:
these should all be hidden from the end user, in fact maybe we should make `Config` and `Recipe` private since this is primarily for us developers to track built-in recipe attributes and their associated configs... an internal registry if you will (@ebsmothers shudders)

Although, I'm not entirely convinced that we really need this. The primary attribute you need to check is `supports_distributed` so you can run with torchrun or runpy on single device. Why the distinction? Why can't we just always use torchrun? And do you anticipate many more recipe attributes like this to be added? Because if we remove `supports_distributed`, we can still do just a simple mapping as we had before which is easier to maintain than the dataclasses

Reply:
I agree that these should all be hidden from the user, but I will push back on the second part.

I originally wanted to run everything through torchrun, but adding on a multiprocessing layer even for simple recipes proved troublesome. 1) It clogs up the logs 2) it was much harder to track down errors even for simple recipes and 3) testing became much more difficult b/c we can't directly observe output (must be logged to file first)

I'm definitely open to revisiting this issue in the future, but it should be noted that it's not a trivial change.

I don't imagine a ton of new recipes to be added to the core library, and I would argue that it's actually easier to maintain this than a separate list and dictionary for recipes and recipes -> configs, which is what we had before.

Reply:
Ok I see your point about distributed logs. That's fair, I can get on board with that